Search all posts and group results into sections

chipsnyder opened the pull request:
Fixes #9183

## Description
This is to handle an enhancement to the current search in the Posts section of the App. Today the app only searches the published posts, where the goal will be to add the other types of posts to the results and mirror the look and feel of  the Pages section of the app.

To do this in `PostListItemDataSource` the app will fetch the next list of data from [GET /sites/$site/posts/](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/posts/) (sorted by date) and them group the posts into the sections. It then inserts the headers (if the section has results) and returns back a list with the grouped elements. 

**Notes:**
- In my testing [GET /sites/$site/posts/](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/posts/) was throwing a random error. Removing "meta" from the fields seems to resolve this in my testing. This seems to be an API issue as I'm seeing it with the existing requests as well. We'll be letting the API team know so we can raise an issue.
- The approach taken here follows the pattern done by WooCommerce but does have the potential for a user to scroll past a section, trigger another page of data, and get updates to a section they passed. Narrowing the results by adding the the search query more will limit this issue. 

## To test:
**Searching Prompt**
Given the user is on the Blog Posts Search screen
Then they should see "Search posts" instead of "Search published posts"

**Searching with results**
Given the user is on the Blog Posts Search screen
When they perform a search with results in multiple statuses
Then they should see the results separated into each status group sorted by date

**Searching with results and scroll**
Given the user is on the Blog Posts Search screen
When they perform a search with results in multiple statuses
And they trigger a new paged to be fetched
Then they should see the sections update with the new pages added.

**Searching with no results (Unchanged behavior)**
Given the user is on the Blog Posts Search screen
When they perform a search with no results
Then they should see a prompt "No posts matching your search"

## PR submission checklist:

- [x] I have considered adding unit tests where possible.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

planarvoid commented on WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt:
```diff
@@ -106,4 +121,54 @@ class PostListItemDataSource(
+        }
+
+        val allItems = mutableListOf<PostListItemIdentifier>()
+        if (listDrafts.isNotEmpty()) {
```
I'd change the order and keep the Published items first 👍

chipsnyder replied:
Oh you're right we mentioned this briefly in our chat. That we'd maintain the order of the tabs instead of the alpha iOS did. I'll flip this.

planarvoid commented on WordPress/src/main/res/values-en-rGB/strings.xml:
```diff
@@ -142,7 +142,7 @@ Language: en_GB
     <string name="post_settings_featured_image_uploading_tap_for_options">Uploading media.\nPlease tap for options.</string>
     <string name="link_open_new_window">Open link in a new window/tab</string>
-    <string name="post_list_search_prompt">Search published posts</string>
+    <string name="post_list_search_prompt">Search posts</string>
```
I think we only have to change the `strings.xml` and the other files get changed automatically in the other languages

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt:
```diff
@@ -127,6 +143,13 @@ class PostListAdapter(
+        private val dividerTitle = itemView.findViewById<TextView>(R.id.divider_text)
+
+        fun onBind(headerItem: SectionHeaderItem) {
+            dividerTitle.text = headerItem.type.toString()
```
I think we should use the `titleResId` on the `type` instead of calling `enum.toString()`

chipsnyder replied:
💯 yup good eye

chipsnyder commented:
@planarvoid This is good to go now. I did make a couple changes I discovered while testing:
1) I forgot to add in the Section header checks in `DiffUtil.ItemCallback` on the adapter so I dropped those in place
2) I realized I didn't need the `isSearch` parameter I added b/c that info was already there in `postListType`
3) Added the EndListIndicator back for fully fetched lists

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt:
```diff
@@ -143,11 +143,11 @@ class PostListAdapter(
 
         fun onBind(headerItem: SectionHeaderItem) {
-            dividerTitle.text = headerItem.type.toString()
+            dividerTitle.text = itemView.context.getString(headerItem.type.titleResId)
```
I think it's cleaner to use directly `dividerTitle.setText(headerItem.type.titleResId)`. The `setText` is overloaded in Java and Kotlin only takes the one with the `String` parameter.

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt:
```diff
@@ -164,6 +190,9 @@ private val PostListDiffItemCallback = object : DiffUtil.ItemCallback<PostListIt
             return oldItem.data == newItem.data
         }
+        if (oldItem is SectionHeaderItem && newItem is SectionHeaderItem) {
+            return oldItem.type == newItem.type
```
I think I'd just compare the full objects here since `SectionHeaderItem` is a data class, it would currently work the same way but it's more future-proof.

chipsnyder replied:
Yeah that's a good point.

planarvoid replied:
hmm, that's really weird since the data class has a default implementation of `equals` but ok, let's leave it as it is 👍

chipsnyder commented:
Updated the set text and left a follow up thought to the is Equal comment. Also I couldn't change the base of this PR so I had to open a new one. I went ahead and opened it here https://github.com/wordpress-mobile/WordPress-Android/pull/10837 so I could take care of any CI issues.

chipsnyder commented:
Closing this in favor of doing one PR to the main repo https://github.com/wordpress-mobile/WordPress-Android/pull/10837#pullrequestreview-322306601
